app/src/main/java/org/schabi/newpipe/player/ui/PlayerUiList.kt

```diff
@@ -0,0 +1,124 @@
+
+import org.schabi.newpipe.util.GuardedByMutex
+import java.util.Optional
+
+class PlayerUiList(vararg initialPlayerUis: PlayerUi) {
+    var playerUis = GuardedByMutex(mutableListOf<PlayerUi>())
+
+    /**
+     * Creates a [PlayerUiList] starting with the provided player uis. The provided player uis
+     * will not be prepared like those passed to [.addAndPrepare], because when
+     * the [PlayerUiList] constructor is called, the player is still not running and it
+     * wouldn't make sense to initialize uis then. Instead the player will initialize them by doing
+     * proper calls to [.call].
+     *
+     * @param initialPlayerUis the player uis this list should start with; the order will be kept
+     */
+    init {
+        playerUis.runWithLockSync {
+            lockData.addAll(listOf(*initialPlayerUis))
+        }
+    }
+
+    /**
+     * Adds the provided player ui to the list and calls on it the initialization functions that
+     * apply based on the current player state. The preparation step needs to be done since when UIs
+     * are removed and re-added, the player will not call e.g. initPlayer again since the exoplayer
+     * is already initialized, but we need to notify the newly built UI that the player is ready
+     * nonetheless.
+     * @param playerUi the player ui to prepare and add to the list; its [PlayerUi.getPlayer]
+     * will be used to query information about the player state
+     */
+    fun addAndPrepare(playerUi: PlayerUi) {
+        if (playerUi.getPlayer().fragmentListener.isPresent) {
+            // make sure UIs know whether a service is connected or not
+            playerUi.onFragmentListenerSet()
+        }
+
+        if (!playerUi.getPlayer().exoPlayerIsNull()) {
+            playerUi.initPlayer()
+            if (playerUi.getPlayer().playQueue != null) {
+                playerUi.initPlayback()
+            }
+        }
+
+        playerUis.runWithLockSync {
+            lockData.add(playerUi)
+        }
+    }
+
+    /**
+     * Destroys all matching player UIs and removes them from the list.
+     * @param playerUiType the class of the player UI to destroy;
+     * the [Class.isInstance] method will be used, so even subclasses will be
+     * destroyed and removed
+     * @param T the class type parameter </T>
+     * */
+    fun <T> destroyAll(playerUiType: Class<T?>) {
+        val toDestroy = mutableListOf<PlayerUi>()
+
+        // short blocking removal from class to prevent interfering from other threads
+        playerUis.runWithLockSync {
+            val new = mutableListOf<PlayerUi>()
+            for (ui in lockData) {
+                if (playerUiType.isInstance(ui)) {
+                    toDestroy.add(ui)
+                } else {
+                    new.add(ui)
+                }
+            }
+            lockData = new
+        }
+        // then actually destroy the UIs
+        for (ui in toDestroy) {
+            ui.destroyPlayer()
+            ui.destroy()
+        }
```

Comment:
In combination with the above change:

```suggestion
    fun <T : PlayerUi> destroyAll(playerUiType: Class<T>) {
        // short blocking removal from class to prevent interfering from other threads
        val (toDestroy, toKeep) = synchronized(playerUis) {
            playerUis.partition { playerUiType.isInstance(it) }
        }
        playerUis.retainAll(toKeep)

        // then actually destroy the UIs
        for (ui in toDestroy) {
            ui.destroyPlayer()
            ui.destroy()
        }
```